src/algebra/category/CommRing/basic.lean

```diff
@@ -14,73 +14,93 @@ universes u v
 
 open category_theory
 
+/-- The category of semirings. -/
+@[reducible] def SemiRing : Type (u+1) := bundled semiring
+
+namespace SemiRing
+
+def of (R : Type u) [semiring R] : SemiRing := bundled.of R
+
+instance (R : SemiRing) : semiring R := R.str
+
+instance bundled_hom : bundled_hom @ring_hom :=
+⟨@ring_hom.to_fun, @ring_hom.id, @ring_hom.comp, @ring_hom.ext⟩
+
+instance has_forget_to_Mon : has_forget SemiRing.{u} Mon.{u} :=
+bundled_hom.mk_has_forget @semiring.to_monoid (λ R₁ R₂ f, f.to_monoid_hom) (λ _ _ _, rfl)
+
+end SemiRing
+
 /-- The category of rings. -/
 @[reducible] def Ring : Type (u+1) := bundled ring
 
-/-- The category of commutative rings. -/
-@[reducible] def CommRing : Type (u+1) := bundled comm_ring
-
 namespace Ring
 
 instance (x : Ring) : ring x := x.str
 
-instance concrete_is_ring_hom : concrete_category @is_ring_hom :=
-⟨by introsI α ia; apply_instance,
-  by introsI α β γ ia ib ic f g hf hg; apply_instance⟩
+def of (R : Type u) [ring R] : Ring := bundled.of R
 
-def of (α : Type u) [ring α] : Ring := ⟨α⟩
+instance bundled_hom : bundled_hom _ :=
+SemiRing.bundled_hom.full_subcategory @ring.to_semiring
 
-abbreviation forget : Ring.{u} ⥤ Type u := forget
-
-instance hom_is_ring_hom {R S : Ring} (f : R ⟶ S) : is_ring_hom (f : R → S) := f.2
+instance has_forget_to_SemiRing : has_forget Ring.{u} SemiRing.{u} :=
+SemiRing.bundled_hom.full_subcategory_has_forget _
 
 end Ring
 
-namespace CommRing
+/-- The category of commutative semirings. -/
+@[reducible] def CommSemiRing : Type (u+1) := bundled comm_semiring
 
-instance (x : CommRing) : comm_ring x := x.str
+namespace CommSemiRing
 
-abbreviation is_comm_ring_hom {α β} [comm_ring α] [comm_ring β] (f : α → β) : Prop :=
-is_ring_hom f
+instance (x : CommSemiRing) : comm_semiring x := x.str
 
-instance concrete_is_comm_ring_hom : concrete_category @is_comm_ring_hom :=
-⟨by introsI α ia; apply_instance,
-  by introsI α β γ ia ib ic f g hf hg; apply_instance⟩
+def of (R : Type u) [comm_semiring R] : CommSemiRing := bundled.of R
 
-def of (α : Type u) [comm_ring α] : CommRing := ⟨α⟩
+instance bundled_hom : bundled_hom _ :=
+SemiRing.bundled_hom.full_subcategory @comm_semiring.to_semiring
 
-abbreviation forget : CommRing.{u} ⥤ Type u := forget
+instance has_forget_to_SemiRing : has_forget CommSemiRing.{u} SemiRing.{u} :=
+bundled_hom.full_subcategory_has_forget _ _
 
-instance hom_is_ring_hom {R S : CommRing} (f : R ⟶ S) : is_ring_hom (f : R → S) := f.2
+/-- The forgetful functor from commutative rings to (multiplicative) commutative monoids. -/
+instance has_forget_to_CommMon : has_forget CommSemiRing.{u} CommMon.{u} :=
+bundled_hom.mk_has_forget
+  @comm_semiring.to_comm_monoid
+  (λ R₁ R₂ f, f.to_monoid_hom)
+  (by intros; refl)
 
-variables {R S T : CommRing.{u}}
+end CommSemiRing
 
--- TODO rename the next two definitions?
-def Int.cast {R : CommRing} : CommRing.of ℤ ⟶ R := { val := int.cast, property := by apply_instance }
```

Comment:
It seems that these definitions got lost in the process... They should've been rewritten in terms of initial objects anyway.
(@semorrison This is another instance where we don't want some abstract big limit thing, but we want a very good handle on this specific initial object. Just as with `prod` discussed elsewhere.)

Reply:
These were needed during earlier constructions of free-forgetful adjunction for CommRing. This works better now, so these constructions aren't needed any more, and are perhaps good riddance!

Reply:
Ok, good